### DISCUSSION TITLE
Prebuilt binaries supporting plugins / custom linters

### DISCUSSION THREAD

**Discussion by ian-h-chamberlain** [golangci/golangci-lint#3361] [General] [answered]
There has been some previous discussion about this in some other issues:

- https://github.com/golangci/golangci-lint/issues/1182
- https://github.com/golangci/golangci-lint/issues/1276#issuecomment-665903858
- https://github.com/golangci/golangci-lint/issues/1276#issuecomment-1304872725

Since some of those weren't totally relevant to the attached issues and/or closed, I thought it might make sense to start a new discussion about ways that it might be possible to use prebuilt `golangci-lint` binaries with plugin support.

As I understand it, there are a few requirements needed to support Go plugins (copy-pasting from https://github.com/golangci/golangci-lint/discussions/3259#discussioncomment-3778846 for reference):

* exact same Go version
* same OS/arch
* same version of shared dependencies
* `CGO_ENABLED=1` during build

I believe (although not 100% positive) that for the first three of these, the Go runtime verifies these and refuses to load a plugin if these conditions are not met. 

I am also under the impression that `CGO_ENABLED=0` is used when building releases to improve portability for the distributed binaries. Although it's not clear to me exactly how, I can understand that there might be two use cases here:
* more portable + no plugins
* less portable + plugins

With that in mind, does it seem like a reasonable option to publish two versions of the binaries, one with plugin support (`CGO_ENABLED=1`) and one without? Existing users could continue to use the same binaries they already use without any change, but custom plugin support (within the constraints listed above) would be possible without requiring a build from source.

----

Barring any functional changes like publishing plugin-compatible binaries, I think there might be two things that would help users understand what's going on with the status quo:

* As mentioned in #1276, if possible, detect configuration for `custom` linters and print a warning/error if the binary was built with `CGO_ENABLED=0` since it will never be able to load those custom linters.
* Update docs in https://golangci-lint.run/contributing/new-linters/#how-to-add-a-private-linter-to-golangci-lint to explicitly mention that building from source with `CGO_ENABLED=1` is required. Currently the docs don't seem to mention building `golangci-lint` _itself_ at all, and only mention building the plugin, which is a bit confusing since really both appear to be required for the custom linter to work.

**Comment by ldez** [comment, marked as answer]
Hello,
it's not possible to pre-build something compatible with plugins because the plugin and the binary must be built with the same Go stack/environment.
`CGO_ENABLED` is only one of the requirements.

**Reply by stevenh** [reply]
Should we endeavour to update the docs and improve the error when this is the case so its clear to the user?

**Reply by ian-h-chamberlain** [reply]
> because the plugin and the binary must be built with the same Go stack/environment.

That means it should still be possible to write a plugin against a specific `golangci-lint` version, right? As long as that `golangci-lint` binary was built with `CGO_ENABLED=1`? 

It seems to me the advantage of having plugin support at all is somewhat diminished if you have to build the binary from scratch anyway — at that point, why not just add your custom lint directly into the binary instead?

**Reply by philiplinell** [reply]
FYI, this is now added to the documentation https://github.com/golangci/golangci-lint/pull/3468.
